### PR TITLE
Update Rust crates

### PR DESCRIPTION
**Pull request by str4d**
This migrates us to the latest revision of the `zcash/librustzcash` crates.

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [x] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Manual tests: Did you update the [manual tests](../blob/main/docs/tests) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [x] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the demo app and try the changes?
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._

# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [x] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after changes and when possible, leave it in a better place._
- [x] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [x] Run the app: Did you run the demo app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt:452]: I do not know if this is the correct way to expose this; given that `seed` is not stored in `Initializer` currently, I figured it was better to not require it there. But I have no idea how these APIs are invoked. I _think_ this is the only actually-public API related to initialization.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/transaction/PersistentTransactionManager.kt
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/transaction/PersistentTransactionManager.kt:150]: I'm still figuring out how these APIs and their overloading work. We can't merge this PR until we fix this overloading, due to @nuttycom's desire to move all of these APIs to use the (intentionally unstable) binary USK encoding (which will make these two APIs have identical signatures).

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt:634]: I added this comment when I thought this was the public API that led to `RustWelding.getShieldedAddress`, but it does not. I am still figuring out how it is populated, and what the access patterns are.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/fb506e7bb7a5c726ac51ee9b8e1475b2d8da61ca..868e7164f230bb011bf13c5faa2f2c59d01c26e4) to add a `SeedRequired` exception and pass the relevant status across the FFI to raise it.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/fb506e7bb7a5c726ac51ee9b8e1475b2d8da61ca..12fd90af0f6ec0d2a3fe9ccf1076efc1f9408f50) to add a `SeedRequired` exception and pass the relevant status across the FFI to raise it.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt
Comment (pacu) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt:1022]: This is related to the layered architecture we've always wanted. It's probably a good discussion to have in terms of kotlin multiplatform

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/transaction/PersistentTransactionManager.kt
Comment (pacu) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/transaction/PersistentTransactionManager.kt:153]: You should just use the USK because these APIs are really internal I don't think that maintaining the String based API will do any good

**Review by pacu — CHANGES_REQUESTED** [review]
I did a first pass over the PR. Tests on my side tell me that it's safe to proceed with USK with total disregard of existing string based API. 

great work!

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt
Comment (str4d) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt:706]: Indeed, but the problem is that then we have two identical APIs named `encode` that don't encode anything, but actually do the important spend authorizing and transaction creating steps for spending and shielding. I don't yet know whether I can just rename them to be more accurate, or whether the overloaded name is depended on in some way.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/transaction/PersistentTransactionManager.kt
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/transaction/PersistentTransactionManager.kt:153]: PersistentTransactionManager is removed with Roomoval, so I wouldn't worry too much about it.  It will be removed from the public API entirely.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt]: We shouldn't change addresses too much just yet; there are changes happening in the Roomoval PR to make these more type safe.

**Comment by ccjernigan** [comment]
I added a followup issue to address how I think the Synchronization factory methods could evolve in the future: https://github.com/zcash/zcash-android-wallet-sdk/issues/725

**Review by ccjernigan — APPROVED** [review]

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Account.kt
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Account.kt:10]: An account must be between `0` and `2^31 - 1` inclusive (`2^31 - 1` is valid for derivation, but is used for legacy `zcashd` purposes, so we may want to exclude that as well for the purposes of the SDK).

Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Account.kt:10]: Kotlin Int is 32-bits.  Effectively it is a Java Integer.  So https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Integer.html#MAX_VALUE

File: sdk-lib/src/androidTest/java/cash/z/ecc/android/sdk/fixture/WalletFixture.kt
Comment (str4d) [inline sdk-lib/src/androidTest/java/cash/z/ecc/android/sdk/fixture/WalletFixture.kt:11]: In my review, I didn't spot where this was used (or a corresponding deletion of this seed phrase from a test). What motivated its addition?

File: CHANGELOG.md
Comment (str4d) [inline CHANGELOG.md:40]: This was accurate as of the commit where I added the `seed` argument, but now that `Initializer` has been removed, it needs to be rewritten.

**Review by str4d — COMMENTED** [review]
utACK f871896c654bfbbde9ea050247b7fa5f9b01c27c. My comments should be addressed before the feature branch is merged, but don't need to block this PR.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Account.kt
Comment (str4d) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Account.kt:10]: Okay, so restricting to positive is equivalent to restricting to a 31-bit range, great. The `2^31 - 1` exclusion may want to happen, but that doesn't have to block this PR.

File: sdk-lib/src/androidTest/java/cash/z/ecc/android/sdk/fixture/WalletFixture.kt
Comment (ccjernigan) [reply sdk-lib/src/androidTest/java/cash/z/ecc/android/sdk/fixture/WalletFixture.kt:11]: We need hard-coded fixture values to write some unit tests, and this is consistent with a pattern we are adopting in the Secant app.

These are used by the unit tests for DerivationTool and the UnifiedSpendingKey.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Account.kt
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Account.kt:10]: I filed a followup issue to come back to this: https://github.com/zcash/zcash-android-wallet-sdk/issues/729

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt]: I updated the comment

File: CHANGELOG.md
Comment (str4d) [inline CHANGELOG.md:41]: This rewrite is insufficient. This constructor's API was completely altered due to the removal of `Initializer`, and the caller now needs to provide up-front several variables that previously they might have been ignoring and leaving to the `Initializer` defaults. The change should be described here (but it doesn't need to include a migration pathway; that can go in the migrations doc later).

**Review by str4d — COMMENTED** [review]
utACK 600e90de6f7574b10400f92a2d6b3f1dbcc4018b modulo changelog comment (I reviewed https://github.com/zcash/zcash-android-wallet-sdk/compare/f871896c654bfbbde9ea050247b7fa5f9b01c27c..600e90de6f7574b10400f92a2d6b3f1dbcc4018b).

File: CHANGELOG.md
Comment (ccjernigan) [reply CHANGELOG.md:41]: Good comments.  I will resolve this later, as the changelog will have a lot more changes with the Roomoval.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt
Comment (pacu) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt:490]: I think the way Initializer is implemented in Android didn't do much. The way I ended up using it was to actually construct other objects and reduce the object graph. On iOS you can pass an Initializer to a CompactBlockProcessor or Synchronizer to get the repositories, etc. Now that I think of the Android SDK didn't do that so this change makes sense in this context.

**Pull request merged**
